### PR TITLE
modify `select_equals_backward` to propage only to a single value

### LINKED ISSUE #35699: backward for tensor.min() and tensor.min(dim=0) behaves differently
## Backward for `tensor.min` behaves differently if `dim` is set 

I noticed that the gradient of the `tensor.min()` function gives a different output when `dim` is set. Namely, when `dim` is set, the gradient will only correspond to one of the entries with the lowest values. If `dim` is not set, then there will be gradient for all entries which have the lowest value

## Code sample to reproduce

```
import torch
a = torch.tensor([0.1, 0.3, 0.1], dtype=torch.float32, requires_grad = True)
a_cp  = torch.tensor([0.1, 0.3, 0.1], dtype=torch.float32, requires_grad = True)
b = a.min()
b.backward()
a.grad  # Output is tensor([1., 0., 1.])
c, d = a_cp.min(dim=0)
c.backward()
a_cp.grad  # Output is tensor([0., 0., 1.])
```

## Problematic for vectorized implementations

This behavior is not expected and it leads to discrepancies in the gradients between vectorized and non implementations. Setting `dim` in `torch.min()` should not change the gradients. 

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen

### PR DESCRIPTION
**Pull request by nikitaved**
Renames `select_equals_backward` to `select_first_equal_backward` and makes sure it propagates to a single value.
Fixes [#35699](https://github.com/pytorch/pytorch/issues/35699).

### CODE REVIEW

**Comment by nikitaved** [comment]
Works for CPU, fails on CUDA in the index_put_ part...

**Comment by nikitaved** [comment]
Will modify `max` once `min` is done.

**Comment by albanD** [comment]
Both min/max/median use that same backward function. So modifying this one function will fix all 3.
Note that we should rename the function to better reflect what it is doing :)

The problem with the Tensor accessor is that you cannot do that for a cuda Tensor as the memory is on the GPU and so only accessible via a kernel.

**Comment by nikitaved** [comment]
Sure, I will rename things once I make it work. How do I detect that I am on CUDA?
For some reason `if (input.device() == at::kCUDA)` does not work :(

**Comment by nikitaved** [comment]
@albanD , what would be the best place for this case to put my code into? Would `test_torch` be alright?

**Comment by nikitaved** [comment]
TODO: update doc, i.e. torch.min() is deterministic, while torch.min(dim=0) is not...

File: test/test_torch.py [hunk 1]
```diff
@@ -315,6 +315,15 @@ def test_max(self):
     def test_min(self):
         self._testSelection(torch.min, min)
 
+    # tests whether min/max/median propagate to a single value
+    def test_select_first_equal_backward(self):
```
Comment (albanD) [inline test/test_torch.py:319]: I think test_autograd.py would be better as you test autograd-related stuff.
Also if you add it [here](https://github.com/pytorch/pytorch/blob/f813e7184e6098de7557a7873442428182f29ad4/test/test_autograd.py#L5220) in the generic tests, you will get the cpu/cuda tests for free by just using the device that is given.

**Comment by albanD** [comment]
The PR looks good to me (just moving the test would make it look a bit better and fix the flake8 errors).
Do you want to do the doc update here or in a follow up PR?

**Comment by nikitaved** [comment]
@albanD , thanks for the review! I changed the tests. I will update the docs in a separate PR.

File: test/test_torch.py (hunk 1 above)
Comment (nikitaved) [reply test/test_torch.py]: Thanks! Done

**Comment by albanD** [comment]
I am a bit surprised that the gradcheck for these ops is not complaining. Is there any special case in the autograd tests that you saw? Or is it just not tested at the moment?

**Comment by nikitaved** [comment]
That is a good question that I asked myself. I did try to have a look, but I am having troubles understanding the test structure. Is there a documentation on this one somewhere?

**Comment by nikitaved** [comment]
I suspect that the tests were generated with unique min values only, because this is the only case where the old code is correct in terms of producing subgradients.

**Comment by albanD** [comment]
I'm afraid there isn't :'( 
Yes I feel like multiple equal values is a point where min() is not differentiable. And so finite differentiation won't work.
I think your tests should be enough to make sure this special case works fine.

**Comment by nikitaved** [comment]
`min` is a function of a single argument. Is true that the finite difference for `min` does `+epsilon` to all the elements of the array?

**Comment by rgommers** [comment]
Looks like this has been good to go for about a month. @albanD is anything else required, or do you think it can be merged?

**Review by albanD — APPROVED** [review]
Sorry it dropped out of my watching list.
Yes this is good to go.

There is most likely going to be a follow up to deal with the perf regression though.

**Comment by nikitaved** [comment]
@albanD , if there is a perf regression, please, let me know, I could pick it up.

File: tools/autograd/templates/Functions.cpp [hunk 2]
```diff
@@ -642,9 +642,20 @@ Tensor _fused_dropout_backward(Tensor grad, Tensor mask, double p1m) {
...
+Tensor select_first_equal_backward(Tensor grad, const Tensor & input, const Tensor & value) {
   auto grad_input = at::zeros_like(input, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
-  grad_input.masked_fill_(input == value, grad);
+
+  // find indices of the first element for which input[idx] == value
+  auto first_value_idx = (input == value).nonzero().select(0, 0);
+
+  // grad_input[first_value_idx]
+  auto grad_input_value = grad_input;
+  for (int64_t dim = 0; dim < grad_input.sizes().size(); ++dim) {
```
Comment (ngimel) [inline tools/autograd/templates/Functions.cpp:653]: nit: grad_input.dim()

File: tools/autograd/templates/Functions.cpp [hunk 3]
```diff
@@ -642,9 +642,20 @@ Tensor _fused_dropout_backward(Tensor grad, Tensor mask, double p1m) {
...
+    grad_input_value = grad_input_value.select(0, first_value_idx[dim].item().to<int64_t>());
+  }
+  // grad_input[first_value_idx] = grad
+  grad_input_value.copy_(grad);
```
Comment (ngimel) [inline tools/autograd/templates/Functions.cpp:657]: wouldn't index_put_ be better here? There should be an incantation that works https://pytorch.org/cppdocs/notes/tensor_indexing.html

File: tools/autograd/templates/Functions.cpp [hunk 4]
```diff
@@ -642,9 +642,20 @@ Tensor _fused_dropout_backward(Tensor grad, Tensor mask, double p1m) {
   }
 }
 
-Tensor select_equals_backward(Tensor grad, const Tensor & input, const Tensor & value) {
+Tensor select_first_equal_backward(Tensor grad, const Tensor & input, const Tensor & value) {
   auto grad_input = at::zeros_like(input, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
```
Comment (ngimel) [inline tools/autograd/templates/Functions.cpp:646]: Talked to @VitalyFedyunin, it's ok to change the format of gradient here to match the format of input (that is, remove LEGACY_CONTIGUOUS argument, zeros_like would do the right thing automatically). It would also make sense to add a test for channels_last/permuted input in this case.

File: tools/autograd/templates/Functions.cpp (hunk 3 above)
Comment (nikitaved) [reply tools/autograd/templates/Functions.cpp:657]: I was thinking about index_put, but how do I split an index tensor into a tuple of tensors?

Comment (ngimel) [reply tools/autograd/templates/Functions.cpp:657]: using `chunk`?

File: tools/autograd/templates/Functions.cpp (hunk 4 above)
Comment (nikitaved) [reply tools/autograd/templates/Functions.cpp:646]: channels_last/permuted? Could you please elaborate, I am not sure I understand :)

Comment (ngimel) [reply tools/autograd/templates/Functions.cpp:646]: If you have a non contiguous input (e.g. created by `torch.randn(2,3).t()`), in most cases you'd like the gradient to have the same strides. Same for channels-last tensor (`torch.randn(2,3,4,5)).contiguous(memory_format=torch.channels_last)`. That could be achieved by creating grad_input with at::zeros_like (w/o an additional argument). You can inspect the gradient that you produce with gradient hooks in the test. Unfortunately just looking at input's .grad attribute won't help because it will be made contiguous (long story, we'll fix it). Also, feel free to ignore it, really it's outside of scope for this PR.

**Review by albanD — APPROVED** [review]
Perfect. Thanks
